### ISSUE TITLE
how to repeat some images when training model

### ISSUE THREAD

**Issue by taofuyu** [ultralytics/ultralytics#22430]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

For example, my train dataset has 100 images, I want to repeat 30 images for 5 times. How to do this?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @taofuyu, thanks for opening this issue and for your interest in Ultralytics 🚀! This is an automated response to help get things moving—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (like dataset sampling or repetition), please provide as much information as possible, including:
-Your exact training command or Python snippet
-Your `data.yaml` and any relevant `cfg`/args you’re using
-Example images/labels that illustrate the scenario
-Training logs and environment details (OS, Python, PyTorch, CUDA)
-And verify you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):
-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You will need to modify the datasets `__len__` and `__getitem__` to return a higher length value and return the indices of the images you want to repeat. You can check an example modification here:

https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Comment by taofuyu** [comment]
> You will need to modify the datasets `__len__` and `__getitem__` to return a higher length value and return the indices of the images you want to repeat. You can check an example modification here:
> 
> https://y-t-g.github.io/tutorials/yolo-class-balancing/

It is too complex, I think there should be a sample way like modify the train.txt or datat.yaml. I repeat some imgs in train.txt, but it brings bad training result.

**Comment by Y-T-G** [comment]
You have very few images. You should have at least 1k images for training. Repeating is not going to help.

**Comment by taofuyu** [comment]
> You have very few images. You should have at least 1k images for training. Repeating is not going to help.

It is just an example

**Comment by glenn-jocher** [comment]
Thanks for clarifying. There isn’t a `data.yaml`/list-file switch to oversample specific images; the simplest supported way is to wrap the dataset in a tiny custom `Trainer` and repeat chosen indices. Example:

```python
from pathlib import Path
import torch
from torch.utils.data import Dataset
from ultralytics import YOLO
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.data.build import build_yolo_dataset, InfiniteDataLoader, seed_worker

class OversampleDataset(Dataset):
    def __init__(self, base, repeat_idx, factor=5):
        self.base = base
        self.map = list(range(len(base))) + repeat_idx * (factor - 1)

    def __len__(self):
        return len(self.map)

    def __getitem__(self, i):
        return self.base[self.map[i]]

    @property
    def collate_fn(self):
        return getattr(self.base, "collate_fn", None)

class OversampleTrainer(DetectionTrainer):
    def get_dataloader(self, dataset_path, batch_size=16, rank=0, mode="train"):
        ds = build_yolo_dataset(self.args, dataset_path, batch_size, self.data, mode=mode, rect=False)
        if mode == "train":
            # pick images to repeat by filename
            names_to_repeat = {"img001.jpg", "img123.jpg"}  # add your 30 names here
            repeat_idx = [i for i, lb in enumerate(ds.labels) if Path(lb["im_file"]).name in names_to_repeat]
            ds = OversampleDataset(ds, repeat_idx=repeat_idx, factor=5)
        return InfiniteDataLoader(ds, batch_size=batch_size, shuffle=(mode == "train"),
                                  num_workers=self.args.workers, pin_memory=True,
                                  worker_init_fn=seed_worker, collate_fn=ds.collate_fn)

model = YOLO("yolo11n.pt")
model.train(data="data.yaml", trainer=OversampleTrainer, imgsz=640, epochs=100)
```

This keeps Ultralytics’ loader intact while repeating your selected images; see the InfiniteDataLoader and dataset builder details in the [InfiniteDataLoader reference](https://docs.ultralytics.com/reference/data/build/) and [build_yolo_dataset reference](https://docs.ultralytics.com/reference/data/build/). If you duplicate paths in a list file, pair it with stronger augmentation to avoid overfitting; our [data augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/) shows safe knobs to increase variability.
